{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>№ п\\п</th>\n",
       "      <th>Диагноз</th>\n",
       "      <th>Тяжесть заболевание</th>\n",
       "      <th>Вегетативно-сосудистая_полинейропатия</th>\n",
       "      <th>Пояснично-крестцовая радикулопатия</th>\n",
       "      <th>Периферический ангиодистонический синдром</th>\n",
       "      <th>Шейно-плечевая радикулопатия</th>\n",
       "      <th>Гипертоническая болезнь</th>\n",
       "      <th>Ишемическая болезнь</th>\n",
       "      <th>ЦВЗ</th>\n",
       "      <th>...</th>\n",
       "      <th>Эритроциты</th>\n",
       "      <th>СОЭ</th>\n",
       "      <th>САД</th>\n",
       "      <th>ДАД</th>\n",
       "      <th>ЧСС</th>\n",
       "      <th>rs4880 гена SOD2</th>\n",
       "      <th>rs361525 гена TNFA</th>\n",
       "      <th>rs16944 гена IL1B</th>\n",
       "      <th>rs1799750 гена MMP1</th>\n",
       "      <th>rs1800795 гена IL6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>VibrationDisease</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>5.60</td>\n",
       "      <td>5</td>\n",
       "      <td>110</td>\n",
       "      <td>70</td>\n",
       "      <td>77</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>VibrationDisease</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>5.19</td>\n",
       "      <td>5</td>\n",
       "      <td>140</td>\n",
       "      <td>90</td>\n",
       "      <td>75</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>VibrationDisease</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>5.20</td>\n",
       "      <td>3</td>\n",
       "      <td>130</td>\n",
       "      <td>80</td>\n",
       "      <td>66</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>VibrationDisease</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>4.71</td>\n",
       "      <td>4</td>\n",
       "      <td>120</td>\n",
       "      <td>85</td>\n",
       "      <td>66</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>VibrationDisease</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>4.75</td>\n",
       "      <td>6</td>\n",
       "      <td>125</td>\n",
       "      <td>80</td>\n",
       "      <td>63</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   № п\\п           Диагноз  Тяжесть заболевание  \\\n",
       "0      1  VibrationDisease                    0   \n",
       "1      2  VibrationDisease                    2   \n",
       "2      3  VibrationDisease                    1   \n",
       "3      4  VibrationDisease                    2   \n",
       "4      5  VibrationDisease                    1   \n",
       "\n",
       "   Вегетативно-сосудистая_полинейропатия  Пояснично-крестцовая радикулопатия  \\\n",
       "0                                   True                               False   \n",
       "1                                  False                               False   \n",
       "2                                  False                               False   \n",
       "3                                  False                               False   \n",
       "4                                   True                               False   \n",
       "\n",
       "   Периферический ангиодистонический синдром  Шейно-плечевая радикулопатия  \\\n",
       "0                                      False                         False   \n",
       "1                                      False                         False   \n",
       "2                                      False                         False   \n",
       "3                                       True                         False   \n",
       "4                                       True                         False   \n",
       "\n",
       "   Гипертоническая болезнь  Ишемическая болезнь    ЦВЗ  ...  Эритроциты  СОЭ  \\\n",
       "0                     True                False  False  ...        5.60    5   \n",
       "1                     True                False  False  ...        5.19    5   \n",
       "2                    False                False  False  ...        5.20    3   \n",
       "3                     True                 True  False  ...        4.71    4   \n",
       "4                     True                False  False  ...        4.75    6   \n",
       "\n",
       "   САД  ДАД  ЧСС  rs4880 гена SOD2  rs361525 гена TNFA  rs16944 гена IL1B  \\\n",
       "0  110   70   77                 3                   2                  2   \n",
       "1  140   90   75                 2                   3                  2   \n",
       "2  130   80   66                 1                   3                  3   \n",
       "3  120   85   66                 2                   3                  1   \n",
       "4  125   80   63                 3                   3                  1   \n",
       "\n",
       "   rs1799750 гена MMP1  rs1800795 гена IL6  \n",
       "0                    1                   1  \n",
       "1                    1                   1  \n",
       "2                    1                   1  \n",
       "3                    2                   1  \n",
       "4                    2                   1  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from scipy.stats import chi2_contingency\n",
    "import numpy as np\n",
    "from scipy.stats.contingency import odds_ratio\n",
    "\n",
    "# Путь к файлу\n",
    "file_path = 'ЭР.xlsm'\n",
    "\n",
    "# Чтение данных из файла Excel\n",
    "try:\n",
    "    # Попытка загрузить файл\n",
    "    genetic_data = pd.read_excel(file_path)\n",
    "except Exception as e:\n",
    "    # В случае ошибки при чтении файла, сохранить сообщение об ошибке\n",
    "    error_message = str(e)\n",
    "    genetic_data = None\n",
    "\n",
    "# Вывод первых нескольких строк данных для проверки содержимого или сообщения об ошибке\n",
    "genetic_data.head() if genetic_data is not None else error_message"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['№ п\\п', 'Диагноз', 'Тяжесть заболевание',\n",
       "       'Вегетативно-сосудистая_полинейропатия',\n",
       "       'Пояснично-крестцовая радикулопатия',\n",
       "       'Периферический ангиодистонический синдром',\n",
       "       'Шейно-плечевая радикулопатия', 'Гипертоническая болезнь',\n",
       "       'Ишемическая болезнь', 'ЦВЗ', 'Остеоартроз', 'БА', 'ХОБЛ',\n",
       "       'Нейросенсорная тугоухость', 'СД', 'Пептическая язва', 'Лейкоциты',\n",
       "       'Эритроциты', 'СОЭ', 'САД', 'ДАД', 'ЧСС', 'rs4880 гена SOD2',\n",
       "       'rs361525 гена TNFA', 'rs16944 гена IL1B', 'rs1799750 гена MMP1',\n",
       "       'rs1800795 гена IL6'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs4880 гена SOD2</th>\n",
       "      <th>rs361525 гена TNFA</th>\n",
       "      <th>rs16944 гена IL1B</th>\n",
       "      <th>rs1799750 гена MMP1</th>\n",
       "      <th>rs1800795 гена IL6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TT</td>\n",
       "      <td>GA</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TC</td>\n",
       "      <td>GG</td>\n",
       "      <td>GA</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TC</td>\n",
       "      <td>GG</td>\n",
       "      <td>AA</td>\n",
       "      <td>Gd</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>AA</td>\n",
       "      <td>Gd</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  rs4880 гена SOD2 rs361525 гена TNFA rs16944 гена IL1B rs1799750 гена MMP1  \\\n",
       "0               TT                 GA                GA                  GG   \n",
       "1               TC                 GG                GA                  GG   \n",
       "2               CC                 GG                GG                  GG   \n",
       "3               TC                 GG                AA                  Gd   \n",
       "4               TT                 GG                AA                  Gd   \n",
       "\n",
       "  rs1800795 гена IL6  \n",
       "0                 CC  \n",
       "1                 CC  \n",
       "2                 CC  \n",
       "3                 CC  \n",
       "4                 CC  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Задаем соответствия для каждого столбца\n",
    "mapping_rs4880 = {1: 'CC', 2: 'TC', 3: 'TT'}\n",
    "mapping_rs361525 = {1: 'AA', 2: 'GA', 3: 'GG'}\n",
    "mapping_rs16944 = {1: 'AA', 2: 'GA', 3: 'GG'}\n",
    "mapping_rs1799750 = {1: 'GG', 2: 'Gd', 3: 'dd'}\n",
    "mapping_rs1800795 = {1: 'CC', 2: 'CG', 3: 'GG'}\n",
    "\n",
    "# Применяем соответствия к данным\n",
    "genetic_data['rs4880 гена SOD2'] = genetic_data['rs4880 гена SOD2'].map(mapping_rs4880)\n",
    "genetic_data['rs361525 гена TNFA'] = genetic_data['rs361525 гена TNFA'].map(mapping_rs361525)\n",
    "genetic_data['rs16944 гена IL1B'] = genetic_data['rs16944 гена IL1B'].map(mapping_rs16944)\n",
    "genetic_data['rs1799750 гена MMP1'] = genetic_data['rs1799750 гена MMP1'].map(mapping_rs1799750)\n",
    "genetic_data['rs1800795 гена IL6'] = genetic_data['rs1800795 гена IL6'].map(mapping_rs1800795)\n",
    "\n",
    "# Проверяем результат преобразования, выводя первые несколько строк обновленного датафрейма\n",
    "genetic_data[['rs4880 гена SOD2', 'rs361525 гена TNFA', 'rs16944 гена IL1B', 'rs1799750 гена MMP1', 'rs1800795 гена IL6']].head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "def odds_ratio_ci(table, alpha=0.05):\n",
    "    # Расчет отношения шансов\n",
    "    or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
    "    \n",
    "    # Расчет стандартной ошибки логарифма отношения шансов\n",
    "    se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
    "    \n",
    "    # Доверительный интервал для логарифма отношения шансов\n",
    "    ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
    "    \n",
    "    # Возврат к оригинальному масштабу отношения шансов\n",
    "    ci_or = [round((np.exp(ci_log_or[0])),2), round((np.exp(ci_log_or[1])),2)]\n",
    "    \n",
    "    return or_value, ci_or\n",
    "\n",
    "def calculate_extended_statistics(genetic_data, genotype_col, feature_col):\n",
    "    # Получаем уникальные генотипы\n",
    "    genotypes = genetic_data[genotype_col].unique()\n",
    "    # Общее количество случаев\n",
    "    total_count = genetic_data.shape[0]\n",
    "    \n",
    "    # Результаты для каждого генотипа\n",
    "    results = []\n",
    "\n",
    "    for genotype in genotypes:\n",
    "        # Таблица сопряженности для данного генотипа\n",
    "        feature_positive = genetic_data[genetic_data[feature_col] == True][genotype_col].value_counts()\n",
    "        feature_negative = genetic_data[genetic_data[feature_col] == False][genotype_col].value_counts()\n",
    "        \n",
    "        \n",
    "        # Количество случаев с признаком и без него для данного генотипа\n",
    "        count_positive = feature_positive.get(genotype, 0)\n",
    "        count_negative = feature_negative.get(genotype, 0)\n",
    "        \n",
    "        # Проценты для каждого генотипа\n",
    "        percent_positive = round(((count_positive / feature_positive.sum()) * 100),2)\n",
    "        percent_negative = round(((count_negative / feature_negative.sum()) * 100), 2)\n",
    "        \n",
    "        # Создание таблицы сопряженности\n",
    "        contingency_table = np.array([[count_positive, count_negative],\n",
    "                                      [feature_positive.sum() - count_positive, \n",
    "                                       feature_negative.sum() - count_negative]])\n",
    "        \n",
    "        # Расчет Хи-квадрат и p-значения\n",
    "        chi2, p, _, _ = chi2_contingency(contingency_table, correction=True)\n",
    "        chi2 = round(chi2, 2)\n",
    "        p = round(p, 2)\n",
    "        # Расчет OR и 95% CI\n",
    "        or_value, ci_or = odds_ratio_ci(contingency_table)\n",
    "        \n",
    "        # Добавление результатов в список\n",
    "        results.append({\n",
    "            'Genotype': genotype,\n",
    "            'With_Feature': count_positive,\n",
    "            'With_Feature_Percent': percent_positive,\n",
    "            'Without_Feature': count_negative,\n",
    "            'Without_Feature_Percent': percent_negative,\n",
    "            'Chi2': chi2,\n",
    "            'p-value': p,\n",
    "            'Odds Ratio': or_value,\n",
    "            '95% CI': ci_or\n",
    "        })\n",
    "    \n",
    "    # Создание DataFrame из результатов\n",
    "    results_df = pd.DataFrame(results)\n",
    "    \n",
    "    return results_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "from docx import Document\n",
    "from docx.oxml.ns import nsdecls\n",
    "from docx.oxml import parse_xml\n",
    "\n",
    "def round_confidence_interval(ci):\n",
    "    return (round(ci[0], 2), round(ci[1], 2))\n",
    "\n",
    "def create_description(statistics_df, feature_name):\n",
    "    descriptions = []\n",
    "    for _, row in statistics_df.iterrows():\n",
    "        genotype = row['Genotype']\n",
    "        with_feature_percent = row['With_Feature_Percent']\n",
    "        without_feature_percent = row['Without_Feature_Percent']\n",
    "        p_value = row['p-value']\n",
    "        orf = row['Odds Ratio']\n",
    "        ci = round_confidence_interval(row['95% CI'])\n",
    "        description = (\n",
    "            f\"В группе больных с признаком {feature_name} генотип '{genotype}' \"\n",
    "            f\"встречается с частотой {with_feature_percent:.2f}%, что \"\n",
    "        )\n",
    "        if p_value < 0.05:\n",
    "            description += (\n",
    "                f\"статистически значимо чаще (p={p_value:.3f}), чем в группе без признака, \"\n",
    "                f\"где этот генотип встречается с частотой {without_feature_percent:.2f}%. \"\n",
    "                f\"Отношения шансов для генотипа момтавляет: {orf}.\"\n",
    "            )\n",
    "        else:\n",
    "            description += (\n",
    "                f\"статистически не отличается (p={p_value:.3f}) от группы без признака, \"\n",
    "                f\"где этот генотип встречается с частотой {without_feature_percent:.2f}%. \"\n",
    "                f\"Отношения шансов для генотипа момтавляет: {orf}.\"\n",
    "            )\n",
    "        descriptions.append(description)\n",
    "    return \" \".join(descriptions)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_allele_statistics(genetic_data, genotype_col, feature_col):\n",
    "    # Получаем уникальные генотипы и определяем аллели\n",
    "    genotypes = genetic_data[genotype_col].unique()\n",
    "    alleles = set(''.join(genotypes))  # Например, из 'CC', 'CT', 'TT' получаем {'C', 'T'}\n",
    "\n",
    "    # Собираем общие подсчеты по аллелям\n",
    "    allele_counts = {allele: {'with_feature': 0, 'without_feature': 0} for allele in alleles}\n",
    "\n",
    "    for genotype in genotypes:\n",
    "        for allele in alleles:\n",
    "            if allele in genotype:\n",
    "                allele_count = genotype.count(allele)\n",
    "                allele_counts[allele]['with_feature'] += allele_count * genetic_data[(genetic_data[genotype_col] == genotype) & (genetic_data[feature_col] == True)].shape[0]\n",
    "                allele_counts[allele]['without_feature'] += allele_count * genetic_data[(genetic_data[genotype_col] == genotype) & (genetic_data[feature_col] == False)].shape[0]\n",
    "\n",
    "    # Формируем таблицу сопряженности из подсчетов аллелей\n",
    "    contingency_table = np.array([list(counts.values()) for _, counts in allele_counts.items()])\n",
    "    \n",
    "    # Расчет процентов для каждой ячейки таблицы\n",
    "    row_sums = contingency_table.sum(axis=1, keepdims=True)\n",
    "    percentages = (contingency_table / row_sums) * 100\n",
    "\n",
    "    # Расчет Хи-квадрат и p-значения\n",
    "    chi2, p, _, _ = chi2_contingency(contingency_table, correction=False)\n",
    "    chi2 = round(chi2, 2)\n",
    "    p = round(p, 3)\n",
    "\n",
    "    # Расчет OR и 95% CI для каждого аллеля\n",
    "# Расчет OR и 95% CI для каждого аллеля\n",
    "    allele_statistics = []\n",
    "    for i, (allele, counts) in enumerate(allele_counts.items()):\n",
    "        # Для первого аллеля используем таблицу как есть, для второго - обратную\n",
    "        if i == 0:\n",
    "            contingency_table = contingency_table\n",
    "        else:\n",
    "            contingency_table = contingency_table[::-1]\n",
    "\n",
    "        or_value, ci_or = odds_ratio_ci(contingency_table)\n",
    "        allele_statistics.append({\n",
    "            'Allele': allele,\n",
    "            'With_Feature': counts['with_feature'],\n",
    "            'With_Feature_Percent': round(counts['with_feature'] / (contingency_table[0,0] + contingency_table[1,0])*100,2), \n",
    "            'Without_Feature': counts['without_feature'],\n",
    "            'Without_Feature_Percent': round(counts['without_feature'] / (contingency_table[0,1] + contingency_table[1,1])*100,2),\n",
    "            'Chi2': chi2,\n",
    "            'p-value': p,\n",
    "            'Odds Ratio': or_value,\n",
    "            '95% CI': ci_or\n",
    "        })\n",
    "\n",
    "\n",
    "    # Создание DataFrame из результатов\n",
    "    allele_results_df = pd.DataFrame(allele_statistics)\n",
    "    \n",
    "    return allele_results_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_description(statistics_df, gene_name, feature_name, is_allele=False):\n",
    "    descriptions = []\n",
    "    for _, row in statistics_df.iterrows():\n",
    "        identifier = row['Allele'] if is_allele else row['Genotype']\n",
    "        with_feature_percent = row['With_Feature_Percent']\n",
    "        without_feature_percent = row['Without_Feature_Percent']\n",
    "        p_value = row['p-value']\n",
    "        orf = row['Odds Ratio']\n",
    "        ci = round_confidence_interval(row['95% CI'])\n",
    "        \n",
    "        # Текст описания в зависимости от того, работаем ли мы с генотипом или аллелем\n",
    "        entity_type = \"аллель\" if is_allele else \"генотип\"\n",
    "        description = (\n",
    "            f\"В группе больных с признаком {feature_name} {entity_type} {identifier} полиморфного локуса {gene_name} \"\n",
    "            f\"встречается с частотой {with_feature_percent:.2f}%, что \"\n",
    "        )\n",
    "        if p_value < 0.05:\n",
    "            description += (\n",
    "                f\"статистически значимо чаще (p={p_value:.3f}), чем в группе без признака, \"\n",
    "                f\"где этот {entity_type} встречается с частотой {without_feature_percent:.2f}%. \"\n",
    "                f\"Отношение шансов для {entity_type}а составляет: {orf}.\"\n",
    "            )\n",
    "        else:\n",
    "            description += (\n",
    "                f\"статистически не отличается (p={p_value:.3f}) от группы без признака, \"\n",
    "                f\"где этот {entity_type} встречается с частотой {without_feature_percent:.2f}%. \"\n",
    "                f\"Отношение шансов для {entity_type}а составляет: {orf}.\"\n",
    "            )\n",
    "        descriptions.append(description)\n",
    "    return \" \".join(descriptions)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in longlong_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in long_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in long_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in long_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in long_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in long_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in long_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in long_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in long_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in long_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in long_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in long_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in long_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Документ успешно сохранен как: genetic_analysis_report.docx\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:6: RuntimeWarning: divide by zero encountered in long_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:9: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\3024606394.py:3: RuntimeWarning: divide by zero encountered in long_scalars\n",
      "  or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n"
     ]
    }
   ],
   "source": [
    "doc = Document()\n",
    "\n",
    "# Список всех признаков для анализа\n",
    "features = [\n",
    "    'Вегетативно-сосудистая_полинейропатия', 'Пояснично-крестцовая радикулопатия', \n",
    "    'Периферический ангиодистонический синдром', 'Шейно-плечевая радикулопатия', \n",
    "    'Гипертоническая болезнь', 'Ишемическая болезнь', 'ЦВЗ', \n",
    "    # 'Остеоартроз', \n",
    "    'БА', 'ХОБЛ', 'Нейросенсорная тугоухость', 'СД', 'Пептическая язва'\n",
    "]\n",
    "\n",
    "# Список генов для анализа\n",
    "genes = ['rs4880 гена SOD2', 'rs361525 гена TNFA', 'rs16944 гена IL1B', 'rs1799750 гена MMP1', 'rs1800795 гена IL6']\n",
    "all_features_results = {}\n",
    "for gene in genes:\n",
    "    gene_results = {}\n",
    "    for feature in features:\n",
    "        feature_statistics = calculate_extended_statistics(genetic_data, gene, feature)\n",
    "        gene_results[feature] = feature_statistics\n",
    "    all_features_results[gene] = gene_results\n",
    "\n",
    "for feature in features:\n",
    "    for gene in genes:\n",
    "        statistics_df = all_features_results[gene][feature]\n",
    "\n",
    "        for col in ['With_Feature', 'Without_Feature', 'With_Feature_Percent', 'Without_Feature_Percent']:\n",
    "            statistics_df[col] = statistics_df[col].apply(lambda x: round(x, 2) if isinstance(x, (int, float)) else x)\n",
    "\n",
    "        statistics_df['p-value'] = statistics_df['p-value'].apply(lambda x: round(x, 3) if isinstance(x, (int, float)) else x)\n",
    "        statistics_df['95% CI'] = statistics_df['95% CI'].apply(round_confidence_interval)\n",
    "\n",
    "\n",
    "        doc.add_paragraph(f\"Таблица 1. \\n Частоты генотипов и аллелей полиморфного локуса {gene} в группах с {feature} и с отсуствием данного признака \")\n",
    "                # Добавляем заголовок перед описанием\n",
    "\n",
    "        # Второй шаг - расчет статистики для аллелей\n",
    "        allele_statistics = calculate_allele_statistics(genetic_data, gene, feature)\n",
    "        allele_descriptions = create_description(allele_statistics, gene,feature, is_allele=True)\n",
    "\n",
    "       \n",
    "        allele_statistics.rename(columns={'Allele': 'Genotype'}, inplace=True)\n",
    "        combined_df = pd.concat([statistics_df, allele_statistics], ignore_index=True)\n",
    "\n",
    "        table = doc.add_table(rows=1, cols=len(combined_df.columns))\n",
    "        table.style = 'Table Grid'\n",
    "        tbl_borders = {\n",
    "            \"sz\": 6, \"val\": \"single\", \"color\": \"000000\", \"space\": 1\n",
    "        }\n",
    "        tbl_xml = parse_xml(r'<w:tblPr {}><w:tblBorders><w:top {}/><w:left {}/><w:bottom {}/><w:right {}/><w:insideH {}/><w:insideV {}/></w:tblBorders></w:tblPr>'.format(\n",
    "            nsdecls('w'), *[f'{k}=\"{v}\"' for k, v in tbl_borders.items()] * 6))\n",
    "        table._element.insert(0, tbl_xml)\n",
    "\n",
    "        hdr_cells = table.rows[0].cells\n",
    "        hdr_titles = [\"Генотип/Аллель\", f\"Наличие {feature}\", \"Процент\", f\"Отсутствие {feature}\", \"Процент\", \"Х2\", \"p\", \"OR\", \"CI95%\"]\n",
    "        for i, title in enumerate(hdr_titles):\n",
    "            hdr_cells[i].text = title\n",
    "\n",
    "        for _, row in combined_df.iterrows():\n",
    "            row_cells = table.add_row().cells\n",
    "            for i, value in enumerate(row):\n",
    "                row_cells[i].text = str(value)\n",
    "\n",
    "        doc.add_paragraph()\n",
    "\n",
    "        doc.add_paragraph(f\"Перепиши в научном стиле. используя красивые и умные слова. и используй разнообразный стиль изложения, чтобы было не шаблонно. не меняй цифры. в конце дай небольшую интерпритацию: \")\n",
    "        description = create_description(statistics_df, gene, feature)\n",
    "        doc.add_paragraph(description)\n",
    "        # Для аллелей\n",
    "\n",
    "        doc.add_paragraph(allele_descriptions)\n",
    "\n",
    "try:\n",
    "    docx_path = 'genetic_analysis_report.docx'\n",
    "    doc.save(docx_path)\n",
    "    success = True\n",
    "except Exception as e:\n",
    "    error_message = str(e)\n",
    "    success = False\n",
    "    docx_path = None\n",
    "\n",
    "if success:\n",
    "    print(f\"Документ успешно сохранен как: {docx_path}\")\n",
    "else:\n",
    "    print(f\"Ошибка при сохранении документа: {error_message}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\1425290231.py:59: RuntimeWarning: divide by zero encountered in long_scalars\n",
      "  se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\1425290231.py:62: RuntimeWarning: divide by zero encountered in log\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
      "C:\\Users\\KarimovDO\\AppData\\Local\\Temp\\ipykernel_3504\\1425290231.py:62: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n"
     ]
    }
   ],
   "source": [
    "def calculate_extended_statistics(genetic_data, genotype_col, feature_col):\n",
    "    # Получаем уникальные генотипы\n",
    "    genotypes = genetic_data[genotype_col].unique()\n",
    "    # Уникальные значения предиктора\n",
    "    feature_values = genetic_data[feature_col].unique()\n",
    "\n",
    "    # Результаты для каждого генотипа\n",
    "    results = []\n",
    "\n",
    "    for genotype in genotypes:\n",
    "        for feature_value in feature_values:\n",
    "            # Подсчет случаев с данным генотипом и уровнем предиктора\n",
    "            count_feature_value = genetic_data[(genetic_data[genotype_col] == genotype) & (genetic_data[feature_col] == feature_value)].shape[0]\n",
    "            \n",
    "            # Подсчет случаев с данным генотипом и другими уровнями предиктора\n",
    "            count_other = genetic_data[(genetic_data[genotype_col] == genotype) & (genetic_data[feature_col] != feature_value)].shape[0]\n",
    "\n",
    "            # Проценты для каждого уровня предиктора\n",
    "            total_count_feature = genetic_data[genetic_data[feature_col] == feature_value].shape[0]\n",
    "            percent_feature_value = round((count_feature_value / total_count_feature) * 100, 2)\n",
    "\n",
    "            # Создание таблицы сопряженности\n",
    "            contingency_table = np.array([[count_feature_value, count_other],\n",
    "                                          [total_count_feature - count_feature_value, \n",
    "                                           genetic_data.shape[0] - total_count_feature - count_other]])\n",
    "            \n",
    "            # Расчет Хи-квадрат и p-значения\n",
    "            chi2, p, _, _ = chi2_contingency(contingency_table, correction=True)\n",
    "            # print(contingency_table)\n",
    "            chi2 = round(chi2, 2)\n",
    "            p = round(p, 2)\n",
    "\n",
    "            # Расчет OR и 95% CI\n",
    "            or_value, ci_or = odds_ratio_ci(contingency_table)\n",
    "            \n",
    "            # Добавление результатов в список\n",
    "            results.append({\n",
    "                'Genotype': genotype,\n",
    "                'Feature_Value': feature_value,\n",
    "                'With_Feature': count_feature_value,\n",
    "                'With_Feature_Percent': percent_feature_value,\n",
    "                'Chi2': chi2,\n",
    "                'p-value': p,\n",
    "                'Odds Ratio': or_value,\n",
    "                '95% CI': ci_or\n",
    "            })\n",
    "\n",
    "    # Создание DataFrame из результатов\n",
    "    results_df = pd.DataFrame(results)\n",
    "    \n",
    "    return results_df\n",
    "\n",
    "\n",
    "def odds_ratio_ci(table, alpha=0.05):\n",
    "    # Расчет отношения шансов\n",
    "    or_value = round((table[0, 0] * table[1, 1]) / (table[0, 1] * table[1, 0]), 2)\n",
    "    \n",
    "    # Расчет стандартной ошибки логарифма отношения шансов\n",
    "    se_log_or = np.sqrt(1/table[0, 0] + 1/table[0, 1] + 1/table[1, 0] + 1/table[1, 1])\n",
    "    \n",
    "    # Доверительный интервал для логарифма отношения шансов\n",
    "    ci_log_or = [np.log(or_value) - se_log_or * 1.96, np.log(or_value) + se_log_or * 1.96]\n",
    "    \n",
    "    # Возврат к оригинальному масштабу отношения шансов\n",
    "    ci_or = [round((np.exp(ci_log_or[0])),2), round((np.exp(ci_log_or[1])),2)]\n",
    "    \n",
    "    return or_value, ci_or\n",
    "\n",
    "\n",
    "def describe_row(row):\n",
    "    genotype = row['Genotype']\n",
    "    feature_value = row['Feature_Value']\n",
    "    count = row['With_Feature']\n",
    "    percent = row['With_Feature_Percent']\n",
    "    p_value = row['p-value']\n",
    "    odds_ratio = row['Odds Ratio']\n",
    "    ci = row['95% CI']\n",
    "\n",
    "    description = (\n",
    "        f\"Перепиши в научном стиле. используя красивые и умные слова. и используй разнообразный стиль изложения, чтобы было не шаблонно. не меняй цифры. в конце дай небольшую интерпритацию: Количество пациентов с генотипом {genotype} и тяжестью заболевания {feature_value} было {count} индивидов, \"\n",
    "        f\"что составило {percent}% всех пациентов с этой тяжестью заболевания. \"\n",
    "    )\n",
    "\n",
    "    if p_value < 0.05:\n",
    "        description += f\"Это статистически значимо отличается от противоположной группы с другой тяжестью заболевания (p={p_value}), \"\n",
    "    else:\n",
    "        description += f\"Что статистически не отличается от противоположной группы с другой тяжестью заболевания (p={p_value}), \"\n",
    "\n",
    "    description += f\"отношение шансов составило OR={odds_ratio} {ci}.\"\n",
    "\n",
    "    return description\n",
    "\n",
    "\n",
    "def export_to_docx(all_features_results, filename=\"combined_analysis_report.docx\"):\n",
    "    # Создаем новый документ\n",
    "    doc = Document()\n",
    "\n",
    "    # Добавляем заголовок\n",
    "    doc.add_heading('Общий Генетический Анализ', level=1)\n",
    "\n",
    "    for gene, features_results in all_features_results.items():\n",
    "        for feature, dataframe in features_results.items():\n",
    "            # Добавляем подзаголовок для гена и характеристики\n",
    "            doc.add_heading(f'{gene} - {feature}', level=2)\n",
    "\n",
    "            # Добавляем таблицу с результатами\n",
    "            table = doc.add_table(rows=1, cols=dataframe.shape[1])\n",
    "            table.style = 'Table Grid'\n",
    "\n",
    "            # Заполняем заголовки таблицы\n",
    "            hdr_cells = table.rows[0].cells\n",
    "            for i, col_name in enumerate(dataframe.columns):\n",
    "                hdr_cells[i].text = col_name\n",
    "\n",
    "            # Добавляем строки таблицы\n",
    "            for index, row in dataframe.iterrows():\n",
    "                row_cells = table.add_row().cells\n",
    "                for i, value in enumerate(row):\n",
    "                    row_cells[i].text = str(value)\n",
    "\n",
    "            # Добавляем описания\n",
    "            descriptions = dataframe.apply(describe_row, axis=1)\n",
    "            for desc in descriptions:\n",
    "                doc.add_paragraph(desc)\n",
    "\n",
    "            # Добавляем разрыв страницы перед следующей комбинацией ген-характеристика\n",
    "            doc.add_page_break()\n",
    "\n",
    "    # Сохраняем документ\n",
    "    doc.save(filename)\n",
    "\n",
    "\n",
    "features = [\n",
    "    'Тяжесть заболевание'\n",
    "]\n",
    "# Список генов для анализа\n",
    "genes = ['rs4880 гена SOD2', 'rs361525 гена TNFA', 'rs16944 гена IL1B', 'rs1799750 гена MMP1', 'rs1800795 гена IL6']\n",
    "all_features_results = {}\n",
    "for gene in genes:\n",
    "    gene_results = {}\n",
    "    for feature in features:\n",
    "        feature_statistics = calculate_extended_statistics(genetic_data, gene, feature)\n",
    "        gene_results[feature] = feature_statistics\n",
    "    all_features_results[gene] = gene_results\n",
    "    # Применение функции к каждой строке DataFrame\n",
    "\n",
    "\n",
    "\n",
    "# Пример использования\n",
    "export_to_docx(all_features_results)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scipy.stats as stats\n",
    "\n",
    "# Предположим, что у вас есть DataFrame с данными\n",
    "# data = ...\n",
    "\n",
    "genes = ['rs4880 гена SOD2', 'rs361525 гена TNFA', 'rs16944 гена IL1B', 'rs1799750 гена MMP1', 'rs1800795 гена IL6']\n",
    "parameters = ['Лейкоциты', 'Эритроциты', 'СОЭ', 'САД', 'ДАД', 'ЧСС']\n",
    "\n",
    "# Функция для расчета статистики по генотипу и аллелю\n",
    "def calculate_statistics(data, gene, parameter):\n",
    "    results = []\n",
    "\n",
    "    # Статистика по генотипам\n",
    "    for genotype in data[gene].unique():\n",
    "        subset = data[data[gene] == genotype]\n",
    "        mean = subset[parameter].mean()\n",
    "        std = subset[parameter].std()\n",
    "        n = len(subset)\n",
    "\n",
    "        results.append({'Type': 'Генотип', 'Category': genotype, 'N': n, 'Среднее': f'{mean:.1f}±{std:.1f}'})\n",
    "    \n",
    "    # ANOVA тест\n",
    "    f_val, p_val = stats.f_oneway(*(data[data[gene] == g][parameter] for g in data[gene].unique()))\n",
    "    results[-1]['F'] = round(f_val, 3)\n",
    "    results[-1]['p-value'] = round(p_val, 3)\n",
    "\n",
    "    # Статистика по аллелям\n",
    "    for allele in set(''.join(data[gene].unique())):\n",
    "        subset = data[data[gene].str.contains(allele)]\n",
    "        mean = subset[parameter].mean()\n",
    "        std = subset[parameter].std()\n",
    "        n = len(subset)\n",
    "\n",
    "        results.append({'Type': 'Аллель', 'Category': allele, 'N': n, 'Среднее': f'{mean:.1f}±{std:.1f}'})\n",
    "    \n",
    "    # t-тест\n",
    "    t_val, p_val = stats.ttest_ind(data[data[gene].str.contains(results[-2]['Category'])][parameter],\n",
    "                                   data[data[gene].str.contains(results[-1]['Category'])][parameter])\n",
    "    results[-1]['t'] = round(t_val, 3)\n",
    "    results[-1]['p-value'] = round(p_val, 3)\n",
    "\n",
    "    return results\n",
    "\n",
    "def describe_statistics(statistics, parameter):\n",
    "    descriptions = []\n",
    "\n",
    "    for item in statistics:\n",
    "        if item['Type'] == 'Генотип':\n",
    "            description = f\"Среднее количество показателя {parameter} в группе индивидов с генотипом {item['Category']} составило {item['Среднее']}\"\n",
    "            descriptions.append(description)\n",
    "        \n",
    "        if 'F' in item:\n",
    "            significance = \"не значимые\" if item['p-value'] >= 0.05 else \"достигли уровня статистической значимости\"\n",
    "            descriptions.append(f\"Различия между группами были {significance} (F = {item['F']}, p= {item['p-value']}).\")\n",
    "\n",
    "    for item in statistics:\n",
    "        if item['Type'] == 'Аллель':\n",
    "            description = f\"В группе с аллелем {item['Category']} среднее значение {parameter} составило {item['Среднее']}\"\n",
    "            descriptions.append(description)\n",
    "        \n",
    "        if 't' in item:\n",
    "            significance = \"не значимо\" if item['p-value'] >= 0.05 else \"значимо\"\n",
    "            direction = \"ниже\" if statistics[-2]['Среднее'] > statistics[-1]['Среднее'] else \"выше\"\n",
    "            descriptions.append(f\"При сравнении средних значений показателя {parameter}, мы обнаружили, что в группе с аллелем {item['Category']} оно было статистически {significance} {direction} (t = {item['t']}, p = {item['p-value']})\")\n",
    "\n",
    "    return ' '.join(descriptions)\n",
    "\n",
    "def create_tables(statistics):\n",
    "    # Создаем отдельные списки для генотипов и аллелей\n",
    "    genotype_data = [item for item in statistics if item['Type'] == 'Генотип']\n",
    "    allele_data = [item for item in statistics if item['Type'] == 'Аллель']\n",
    "\n",
    "    # Создаем DataFrame для генотипов\n",
    "    genotype_df = pd.DataFrame(genotype_data)\n",
    "    genotype_df = genotype_df.drop(['Type'], axis=1)\n",
    "\n",
    "    # Создаем DataFrame для аллелей\n",
    "    allele_df = pd.DataFrame(allele_data)\n",
    "    allele_df = allele_df.drop(['Type'], axis=1)\n",
    "\n",
    "    return genotype_df, allele_df\n",
    "\n",
    "\n",
    "# Пример использования\n",
    "all_stats = {}\n",
    "for gene in genes:\n",
    "    gene_stats = {}\n",
    "    for parameter in parameters:\n",
    "        gene_stats[parameter] = calculate_statistics(genetic_data, gene, parameter)\n",
    "    all_stats[gene] = gene_stats\n",
    "\n",
    "from docx import Document\n",
    "from docx.shared import Pt\n",
    "from docx.oxml.ns import qn\n",
    "from docx.enum.text import WD_ALIGN_PARAGRAPH\n",
    "import copy\n",
    "\n",
    "def save_to_docx(all_stats, filename=\"genetic_analysis_report_mean.docx\"):\n",
    "    # Создаем документ\n",
    "    doc = Document()\n",
    "\n",
    "    for gene in genes:\n",
    "        for parameter in parameters:\n",
    "            statistics = all_stats[gene][parameter]\n",
    "            genotype_df, allele_df = create_tables(statistics)\n",
    "            description = describe_statistics(statistics, parameter)\n",
    "\n",
    "            # Добавляем подзаголовок для таблицы генотипов\n",
    "            doc.add_paragraph(f\"Таблица 1. Среднее значение параметра {parameter} и статистическая значимость различий в зависимости от генотипа и наличия аллелей полиморфного локуса {gene}:\", style='Caption')\n",
    "\n",
    "            # Добавляем таблицу генотипов\n",
    "            add_table_to_doc(doc, genotype_df)\n",
    "\n",
    "            # Добавляем таблицу аллелей\n",
    "            add_table_to_doc(doc, allele_df)\n",
    "\n",
    "            # Добавляем описание\n",
    "            p = doc.add_paragraph()\n",
    "            p.add_run(\"Перепиши в научном стиле. используя красивые и умные слова. и используй разнообразный стиль изложения, чтобы было не шаблонно. не меняй цифры. в конце дай небольшую интерпритацию: \").bold = True\n",
    "            p.add_run(description)\n",
    "\n",
    "            # Добавляем разделитель между разными комбинациями\n",
    "            doc.add_paragraph(\"\\n\" + \"-\"*50 + \"\\n\")\n",
    "\n",
    "    # Сохраняем документ\n",
    "    doc.save(filename)\n",
    "\n",
    "def add_table_to_doc(doc, df):\n",
    "    table = doc.add_table(rows=1, cols=len(df.columns))\n",
    "    table.style = 'Table Grid'\n",
    "\n",
    "    # Заполняем заголовки таблицы\n",
    "    hdr_cells = table.rows[0].cells\n",
    "    for i, col_name in enumerate(df.columns):\n",
    "        hdr_cells[i].text = col_name\n",
    "\n",
    "    # Добавляем строки таблицы\n",
    "    for index, row in df.iterrows():\n",
    "        row_cells = table.add_row().cells\n",
    "        for i, value in enumerate(row):\n",
    "            row_cells[i].text = str(value)\n",
    "\n",
    "    # Объединение ячеек\n",
    "    merge_cells(table, 1, 3, 4)  # Объединение ячеек в столбце 'F'\n",
    "    merge_cells(table, 1, 4, 4)  # Объединение ячеек в столбце 'p-value' для генотипов\n",
    "    merge_cells(table, 5, 3, 6)  # Объединение ячеек в столбце 't' для аллелей\n",
    "    merge_cells(table, 5, 4, 6)  # Объединение ячеек в столбце 'p-value' для аллелей\n",
    "\n",
    "def merge_cells(table, start_row, col_idx, end_row):\n",
    "    num_rows = len(table.rows)\n",
    "    # Убедимся, что start_row и end_row находятся в пределах таблицы\n",
    "    start_row = max(1, min(start_row, num_rows - 1))\n",
    "    end_row = min(end_row, num_rows)\n",
    "    table_copy = copy.deepcopy(table)\n",
    "    # Объединяем ячейки\n",
    "    cell_to_merge = table.cell(start_row, col_idx)\n",
    "    for i in range(start_row + 1, end_row):\n",
    "        cell = table.cell(i, col_idx)\n",
    "        cell_to_merge.merge(cell)\n",
    "        # Устанавливаем текст из последней ячейки объединения\n",
    "    cell_to_merge.text = table_copy.cell(end_row - 1, col_idx).text\n",
    "\n",
    "# Пример использования\n",
    "save_to_docx(all_stats)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
